### PR TITLE
FlyDSL sage v1

### PR DESCRIPTION
**Pull request by hellozhuo-amd**
## Summary

This PR adds a FlyDSL implementation of Sage attention for AMD CDNA GPUs (gfx942/gfx950), covering the INT8 Q/K × FP8 V flash-attention path and the fused Q+K+V quantization preprocessor. The kernels match the semantics of the existing Triton `fav3_sage` reference and plug into the same wrapper-level API.

Related Triton reference: `aiter/ops/triton/fav3_sage.py` and bench `op_tests/triton_tests/bench_fav3_sage.py` (refactored in #2920).

## What changed

* **FlyDSL Sage attention kernel (CDNA)**
    * New kernel at `aiter/ops/flydsl/kernels/sage_attn_cdna.py`
    * INT8 Q/K MFMA (32×32×32) for QK, FP8 V MFMA (K=64) for PV on gfx950 with `ds_read_tr8_b64` V LDS transposition; BF16 PV fallback on gfx942
    * Per-element V descaling matching Triton `fav3_sage` semantics
    * GQA, causal masking, BSHD/BHSD layout, unaligned seq-len padding
    * 2-stage software pipeline on K/V LDS; KV bounds-mask fast path; OOB-mask block hoisted out of the inner loop
    * Dynamic `BLOCK_M` selection for occupancy; default `block_n=64`
    * Post-MI scheduling enabled (LLVM post-RA)

* **Fused Q+K+V quantization kernel**
    * New kernel at `aiter/ops/flydsl/kernels/sage_quant_cdna.py`
    * Single-launch fused Q-INT8 + K-INT8 + V-FP8 quantization plus `V_scale` and `K_mean` precompute (gated on `b*h*kv_len` for long-S parity)

* **Wrappers and dispatch**
    * `aiter/ops/flydsl/sage_kernels.py` — top-level Sage attention API with build cache and auto-quantization
    * `aiter/ops/flydsl/sage_quant.py` — sage_quant dispatcher
    * Module exports updated in `aiter/ops/flydsl/__init__.py`

* **Tests and benchmark**
    * `op_tests/flydsl_tests/test_flydsl_sage.py` — correctness vs FP32 SDPA and vs Triton `fav3_sage` reference
    * `op_tests/flydsl_tests/bench_flydsl_sage.py` — FlyDSL vs Triton vs PyTorch SDPA throughput sweep

## Motivation

* Provide a FlyDSL Sage attention path on CDNA so downstream consumers can use the lower-precision (INT8 Q/K, FP8 V) flash-attention flow without taking a Triton dependency on the hot path.
* Land a kernel that is at parity-or-better with the Triton `fav3_sage` reference across the benchmark sweep, so the FlyDSL path can be the default on gfx950 while remaining safe on gfx942.
* Establish the FlyDSL kernel/wrapper layout under `aiter/ops/flydsl/` for future Sage-style kernels (QK_INT8 + V_FP8 + per-block scale).

## Technical Details

* QK uses INT8 32×32×32 MFMA; PV uses K=64 FP8 MFMA on gfx950 with V stored row-major in LDS and read via `ds_read_tr8_b64`. PV falls back to BF16 on gfx942.
* P-handoff between QK and PV uses `v_permlane32_swap_b32` directly on gfx940+ to avoid the generic shuffle path.
* K LDS stride padded to `HEAD_DIM + 16` and V LDS row stride padded by 16 B on gfx950 to avoid bank conflicts.
* Buffer descriptors used for K/V global loads; descriptor tightened to drop the OOB select in the steady-state path.
* Cross-iteration softmax pipeline replaced with a simpler intra-iteration schedule that lets LLVM's post-RA scheduler reach the better issue order.
* Residual scalar muls in the softmax/quant epilogue are packed.
* FP8_MAX P-quant scaling dropped (matches Triton reference numerics).

## Test Plan

Correctness:

```
python3 -m pytest op_tests/flydsl_tests/test_flydsl_sage.py -v
```

Benchmark (note: bench is sensitive to GPU pinning — pin GPU 0 for stable Triton numbers):

```
HIP_VISIBLE_DEVICES=0 python3 op_tests/flydsl_tests/bench_flydsl_sage.py
```

Reference comparison:

```
HIP_VISIBLE_DEVICES=0 python3 op_tests/triton_tests/bench_fav3_sage.py
```

## Test Result

Measured on gfx950 with `HIP_VISIBLE_DEVICES=0 python3 op_tests/flydsl_tests/bench_flydsl_sage.py`.

**Accuracy (cosine similarity vs FP32 SDPA):**

| Shape | FlyDSL cos_min | FlyDSL cos_mean | Triton cos_min |
|---|---|---|---|
| B1 S1024 H8/8 D128 fwd | 0.9986 | 0.9993 | 0.9986 |
| B1 S3000 H8/8 D128 fwd | 0.9984 | 0.9992 | 0.9985 |
| B1 S4096 H8/8 D128 fwd | 0.9984 | 0.9992 | 0.9984 |
| B1 S4096 H8/8 D128 causal | 0.9983 | 0.9993 | Triton FAILED |
| B1 S8192 H8/8 D128 fwd | 0.9982 | 0.9992 | 0.9982 |
| B1 S8192 H8/8 D128 causal | 0.9982 | 0.9992 | Triton FAILED |
| B1 S4096 H16/4 D128 fwd | 0.9984 | 0.9992 | 0.9984 |
| B2 S4096 H16/4 D128 fwd | 0.9984 | 0.9992 | 0.9983 |
| B2 S4096 H8/8 D128 fwd | 0.9982 | 0.9992 | 0.9983 |
| B1 S16384 H8/8 D128 fwd | 0.9979 | 0.9992 | 0.9980 |
| B1 S16384 H24/24 D128 fwd | 0.9974 | 0.9992 | 0.9974 |

**Speed (TFLOPS, vs SDPA / vs Triton):**

| Shape | SDPA TFLOPS | Triton TFLOPS | FlyDSL TFLOPS | vs SDPA | vs Triton |
|---|---|---|---|---|---|
| B1 S1024 H8/8 D128 fwd | 21.06 | 34.54 | 50.37 | 2.39× | **1.46×** |
| B1 S3000 H8/8 D128 fwd | 53.87 | 231.84 | 366.90 | 6.81× | **1.58×** |
| B1 S4096 H8/8 D128 fwd | 70.04 | 365.89 | 548.53 | 7.83× | **1.50×** |
| B1 S4096 H8/8 D128 causal | 38.20 | FAILED | 163.76 | 4.29× | N/A |
| B1 S8192 H8/8 D128 fwd | 97.25 | 832.83 | 901.36 | 9.27× | **1.08×** |
| B1 S8192 H8/8 D128 causal | 56.17 | FAILED | 379.15 | 6.75× | N/A |
| B1 S4096 H16/4 D128 fwd | 71.33 | 713.06 | 863.13 | 12.10× | **1.21×** |
| B2 S4096 H16/4 D128 fwd | 80.59 | 768.20 | 966.38 | 11.99× | **1.26×** |
| B2 S4096 H8/8 D128 fwd | 71.55 | 534.16 | 786.40 | 10.99× | **1.47×** |
| B1 S16384 H8/8 D128 fwd | 99.54 | 1123.90 | 1133.45 | 11.39× | **1.01×** |
| B1 S16384 H24/24 D128 fwd | 99.78 | 1302.61 | 1306.28 | 13.09× | **1.00×** |

**Headlines:**

* All shapes pass correctness (cos_min ≥ 0.9974) against FP32 SDPA, matching Triton parity where Triton runs.
* Triton `fav3_sage` reference **fails on the causal shapes** (S=4096 and S=8192); FlyDSL passes both.
* Geomean speedup vs Triton across the 9 non-causal shapes Triton can run: **~1.27×**.
* Short/medium S (1024–4096): **1.21–1.58× Triton**.
* Long S (16384): **~parity** (1.00–1.01×) — both paths are bandwidth-bound here.
* All FlyDSL shapes are well above SDPA (2.4×–13.1×).

## Notes

* gfx942 falls back to BF16 PV (no `ds_read_tr8_b64`); gfx950 uses the FP8 PV + transposed V LDS path.
* Bench numbers are unstable without `HIP_VISIBLE_DEVICES=0` — pin the GPU when reproducing.

## Submission Checklist

* Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests

### CODE REVIEW

**Comment by zufayu** [comment]
@hellozhuo-amd This PR has been inactive for over 2 months. Is FlyDSL sage v1 still being pursued? If you still plan to pursue it, please rebase onto current main and push an update. If there's no response within ~2 weeks, we'll close this to keep the PR queue clean — it can always be reopened or resubmitted.

**Comment by hellozhuo-amd** [comment]
@zufayu I will close it atm and resubmit it when it's ready.
